Tests/SwiftLexicalScopesTest/SimpleQueryTests.swift

```diff
@@ -0,0 +1,179 @@
+    )
+  }
+
+  func testNoLabeledStatements() {
+    assertLexicalScopeQuery(
+      source: """
+        while true {
+          1️⃣break
+        }
+        """,
+      methodUnderTest: { argument in
+        argument.lookupLabeledStmts()
+      },
+      expected: ["1️⃣": []]
+    )
+  }
+
+  func testLabeledStmtLookupClassNestedWithinLoop() {
+    assertLexicalScopeQuery(
+      source: """
+        1️⃣one: while true {
+          class a {
+            func foo() {
+              2️⃣two: while true {
+                3️⃣break
+              }
+            }
+          }
+          4️⃣break
+        }
+        """,
+      methodUnderTest: { argument in
+        argument.lookupLabeledStmts()
+      },
+      expected: ["3️⃣": ["2️⃣"], "4️⃣": ["1️⃣"]]
+    )
+  }
+
+  func testLabeledStmtLookupClosureNestedWithinLoop() {
+    assertLexicalScopeQuery(
+      source: """
+        1️⃣one: while true {
+          var a = {
+            2️⃣two: while true {
+              3️⃣break
+            }
+          }
+          4️⃣break
+        }
+        """,
+      methodUnderTest: { argument in
+        argument.lookupLabeledStmts()
+      },
+      expected: ["3️⃣": ["2️⃣"], "4️⃣": ["1️⃣"]]
+    )
+  }
+
+  func testLabeledStmtLookupFunctionNestedWithinLoop() {
+    assertLexicalScopeQuery(
+      source: """
+        1️⃣one: while true {
+          func foo() {
+            2️⃣two: while true {
+              3️⃣break
+            }
+          }
+          4️⃣break
+        }
+        """,
+      methodUnderTest: { argument in
+        argument.lookupLabeledStmts()
+      },
+      expected: ["3️⃣": ["2️⃣"], "4️⃣": ["1️⃣"]]
+    )
+  }
+
+  func testLookupFallthroughDestination() {
+    assertLexicalScopeQuery(
+      source: """
+        func foo() {
+          7️⃣print(0)
+        }
+
+        switch a {
+        1️⃣case 1:
+          2️⃣print(1)
+        3️⃣case 2:
+          4️⃣print(2)
+        5️⃣default:
+          6️⃣print(3)
+        }
+        """,
+      methodUnderTest: { argument in
+        let result = argument.lookupFallthroughSourceAndDest()
+        return [result.source, result.destination]
+      },
+      expected: ["2️⃣": ["1️⃣", "3️⃣"], "4️⃣": ["3️⃣", "5️⃣"], "6️⃣": ["5️⃣", nil], "7️⃣": [nil, nil]]
+    )
+  }
+
+  func testLookupCatchNode() {
+    assertLexicalScopeQuery(
+      source: """
+        1️⃣func foo() throws {
+          2️⃣do {
+            try 3️⃣f()
+            4️⃣try? 5️⃣f()
+          } catch {
+            throw 6️⃣error
+          }
+        }
+
+        8️⃣func bar() {
+          throw 7️⃣f()
+        }
+        """,
+      methodUnderTest: { argument in
+        return [argument.lookupCatchNode()]
+      },
+      expected: ["3️⃣": ["2️⃣"], "5️⃣": ["4️⃣"], "6️⃣": ["1️⃣"], "7️⃣": ["8️⃣"]]
+    )
+  }
+
+  func testLookupCatchNodeWithNestedDoCatch() {
+    assertLexicalScopeQuery(
+      source: """
+        1️⃣func foo() rethrows {
+          2️⃣do {
+            3️⃣do {
+              try 4️⃣f()
```

Comment:
It's probably worth testing `try x + y` and `try! x + y`, where we look up from the `y`, to make sure we get the appropriate catch node. We just tripped over this [same issue](https://github.com/swiftlang/swift/pull/74454/commits/8d6434410b591cbf6f8cfc0c729ab27e8aa15ebc) with the C++ implementation .